src/coding_theory/hamming.lean

```diff
@@ -0,0 +1,245 @@
+variables {γ : ι → Type*} [Π i, decidable_eq (γ i)]
+
+/-- The Hamming distance function to the naturals. -/
+def hamm_dist (x y : Π i, β i) := card {i // x i ≠ y i}
+
+lemma hamm_dist_nonneg {x y : Π i, β i} : 0 ≤ hamm_dist x y := zero_le _
+
+lemma hamm_dist_le_card_fintype {x y : Π i, β i} : hamm_dist x y ≤ card ι := card_subtype_le _
+
+lemma hamm_dist_comp_le_hamm_dist (f : Π i, γ i → β i) {x y : Π i, γ i} :
+  hamm_dist (λ i, f i (x i)) (λ i, f i (y i)) ≤ hamm_dist x y :=
+card_subtype_mono _ _ $ λ x H1 H2, H1 $ congr_arg (f x) H2
+
+lemma hamm_dist_comp (f : Π i, γ i → β i) {x y : Π i, γ i} (hf : Π i, injective (f i))
+  : hamm_dist (λ i, f i (x i)) (λ i, f i (y i)) = hamm_dist x y :=
+le_antisymm (hamm_dist_comp_le_hamm_dist  _) (card_subtype_mono _ _ (λ i H1 H2, H1 (hf i H2)))
+
+lemma hamm_dist_smul_le_hamm_dist [Π i, has_scalar α (β i)] {k : α} {x y : Π i, β i} :
+  hamm_dist (k • x) (k • y) ≤ hamm_dist x y :=
+hamm_dist_comp_le_hamm_dist $ λ i, (•) k
+
+lemma hamm_dist_smul [Π i, has_scalar α (β i)] {k : α} {x y : Π i, β i}
+  (hk : Π i, is_smul_regular (β i) k) : hamm_dist (k • x) (k • y) = hamm_dist x y :=
+hamm_dist_comp (λ i, (•) k) hk
+
+@[simp] lemma hamm_dist_eq_zero {x y : Π i, β i} : hamm_dist x y = 0 ↔ x = y :=
+by simp_rw [  function.funext_iff, hamm_dist, card_eq_zero_iff,
+              is_empty_iff, subtype.forall, imp_false, not_not]
+
+@[simp] lemma hamm_dist_self (x : Π i, β i) : hamm_dist x x = 0 := hamm_dist_eq_zero.mpr rfl
+
+lemma hamm_dist_comm (x y : Π i, β i) : hamm_dist x y = hamm_dist y x :=
+by simp_rw [hamm_dist, ne_comm]
+
+lemma hamm_dist_triangle (x y z : Π i, β i) :
+  hamm_dist x z ≤ hamm_dist x y + hamm_dist y z :=
+begin
+  simp_rw hamm_dist,
+  refine le_trans (card_subtype_mono _ _ (λ _ h, _)) (card_subtype_or _ _),
+  by_contra' H, exact h (eq.trans H.1 H.2)
+end
+
+lemma eq_of_hamm_dist_eq_zero (x y : Π i, β i) : hamm_dist x y = 0 → x = y :=
+hamm_dist_eq_zero.mp
+
+lemma hamm_dist_ne_zero {x y : Π i, β i} : hamm_dist x y ≠ 0 ↔ x ≠ y :=
+not_iff_not.mpr hamm_dist_eq_zero
+
+lemma hamm_dist_pos {x y : Π i, β i} : 0 < hamm_dist x y ↔ x ≠ y :=
+by rw [←hamm_dist_ne_zero, iff_not_comm, not_lt, nat.le_zero_iff]
+
+@[simp] lemma hamm_dist_lt_one {x y : Π i, β i} : hamm_dist x y < 1 ↔ x = y :=
+by {rw nat.lt_one_iff, exact hamm_dist_eq_zero}
+
+lemma hamm_dist_eq_zero_iff_forall_eq_zero {x y : Π i, β i} :
+  hamm_dist x y = 0 ↔ ∀ i, x i = y i :=
+by rw [hamm_dist_eq_zero, function.funext_iff]
+
+lemma hamm_dist_ne_zero_iff_exists_ne_zero {x y : Π i, β i} :
+  hamm_dist x y ≠ 0 ↔ ∃ i, x i ≠ y i :=
+by rw [hamm_dist_ne_zero, function.ne_iff]
+
+section has_zero
+
+variables [Π i, has_zero (β i)] [Π i, has_zero (γ i)]
+
+/-- The Hamming weight function to the naturals. -/
+
+def hamm_wt (x : Π i, β i) : ℕ := hamm_dist x 0
+
+lemma hamm_wt_eq (x : Π i, β i) : hamm_wt x = card {i // x i ≠ 0} := rfl
+
+lemma hamm_wt_eq_hamm_dist_zero (x : Π i, β i) : hamm_wt x = hamm_dist x 0 := rfl
```

Comment:
I think you should try to match the `norm` / `dist` API here, as
```suggestion
@[simp] lemma hamm_dist_zero_right (x : Π i, β i) : hamm_dist x 0 = hamm_wt x := rfl

@[simp] lemma hamm_dist_zero_left (x : Π i, β i) : hamm_dist 0 x = hamm_wt x := hamm_dist_comm _ _
```

Reply:
Aha. Are there other changes needed?

Reply:
I'm not sure. Perhaps the best thing to do is add a docstring to every lemma referencing the corresponding lemma for `norm`/`dist`, if one exists.